### PR TITLE
receive: memoize exemplar/TSDB clients

### PR DESCRIPTION
**Pull request by GiedriusS**
We call this on each Series() so memoize the creation of this slice.

### CODE REVIEW

**Review by niaurys — APPROVED** [review]

**Comment by jnyi** [comment]
this pr introduced a bug when multiple tenants are created at the same time, only partially would be memorized due to atomicity issue, fyi @GiedriusS we plan to revert this pr internally but worth taking a look, it prevents new tenants to be queriable but luckily no data loss

File: pkg/receive/multitsdb.go [hunk 1]
```diff
@@ -595,7 +605,17 @@ func (t *MultiTSDB) RemoveLockFilesIfAny() error {
...
+		t.mtx.RUnlock()
+		return t.tsdbClients
+	}
+
+	t.mtx.RUnlock()
+	t.mtx.Lock()
+	if !t.tsdbClientsNeedUpdate {
+		return t.tsdbClients
+	}
+	defer t.mtx.Unlock()
```
Comment (hczhu-db) [inline pkg/receive/multitsdb.go:618]: `defer t.mtx.Unlock()` should be before `if`. @jnyi

Comment (GiedriusS) [reply pkg/receive/multitsdb.go:618]: Thanks; care opening up a PR with the fix?

Comment (GiedriusS) [reply pkg/receive/multitsdb.go:618]: Already fixed here https://github.com/thanos-io/thanos/commit/c10b695a5ba34f05e2328d74dd5e48159137b11e

Comment (jnyi) [reply pkg/receive/multitsdb.go:618]: no, we found a different bug, I submit a PR to fix it, please take a look when you have time: https://github.com/thanos-io/thanos/pull/7941

**Pull request merged**